lib/SILOptimizer/Mandatory/DiagnoseVarUsage.cpp

```diff
@@ -0,0 +1,1617 @@
+  }
+  return nullptr;
+}
+
+//===----------------------------------------------------------------------===//
+//                          MARK: Def-Use Traversers
+//===----------------------------------------------------------------------===//
+
+/// Traverses through the def-use of an instruction representing a var
+/// declaration to find mutations of the value.
+class VariableModifyTraverser {
+  
+  SmallVector<SILValue, 32> Worklist;
+  llvm::SetVector<SILValue> Seen;
+  SILInstruction *EntryInst;
+  
+  bool IsUninitialized = false;
+  bool IsAssigned = false;
+  bool IsWeakCapture = false;
+  
+public:
+  VariableModifyTraverser(SILInstruction *EntryInst) : EntryInst(EntryInst) {
+    appendValues(EntryInst);
+  }
+  
+private:
+  
+  /// Stage instruction results for traversal
+  void appendValues(SILInstruction *I) {
+    for (SILValue Value : I->getResults()) {
+      appendValue(Value);
+    }
+  }
+  
+  /// Stage a value for traversal
+  void appendValue(SILValue Value) {
+    if (!Seen.count(Value)) {
+      Worklist.push_back(Value);
+      Seen.insert(Value);
+    }
+  }
+  
+  /// Determine if external SIL function modifies a value. Returns true if the
+  /// parameter is an inout.
+  static bool externalFunctionArgIsModified(SILInstruction *Apply,
+                                            Operand *Op) {
+    
+    ApplySite AS;
+    SILFunction *Fn;
+    
+    if (auto AI = dyn_cast<ApplyInst>(Apply)) {
+      AS = ApplySite(AI);
+      Fn = AI->getReferencedFunctionOrNull();
+    } else if (auto PAI = dyn_cast<PartialApplyInst>(Apply)) {
+      AS = ApplySite(PAI);
+      Fn = PAI->getReferencedFunctionOrNull();
+    } else if (auto TAI = dyn_cast<TryApplyInst>(Apply)) {
+      AS = ApplySite(TAI);
+      Fn = TAI->getReferencedFunctionOrNull();
+    } else if (auto BAI = dyn_cast<BeginApplyInst>(Apply)) {
+      AS = ApplySite(BAI);
+      Fn = BAI->getReferencedFunctionOrNull();
+    }
+    
+    if (AS && Fn && Fn->isExternalDeclaration()) {
+      if ((Op->getUser() == Apply) && AS.isArgumentOperand(*Op)) {
+        return argIsInout(AS.getArgumentConvention(*Op));
+      }
+    }
+    return false;
+  }
+  
+  // A narrow case for Builtin.convertUnownedUnsafeToGuaranteed
+  //
+  // Check that the order of instructions is:
+  //  - begin_access [read]
+  //  - struct_element_addr
+  //  - load
+  //  - unmanaged_to_ref
+  //
+  bool isConvertUnownedUnsafeToGuaranteed(BeginAccessInst *BAI) {
+    for (SILValue Value : BAI->getResults()) {
+      for (auto Use : Value->getUses()) {
+        auto User = Use->getUser();
+        
+        if (isa<StructElementAddrInst>(User)) {
+          for (SILValue SEAValue : User->getResults()) {
+            for (auto SEAUse : SEAValue->getUses()) {
+              auto SEAUser = SEAUse->getUser();
+              
+              if (isa<LoadInst>(SEAUser)) {
+                for (SILValue LIValue : SEAUser->getResults()) {
+                  for (auto LIUse : LIValue->getUses()) {
```

Comment:
[I don't have a concrete suggestion here at the moment but] The number of nested loops here looks a bit scary. 😰

Reply:
This does look concerning, but I'm not sure how else to check for such a chain of instructions.

A little background here: this is a very weird case for identifying the use of a single Builtin instruction. As far as I can tell, [this will only get triggered once when performing diagnostics within the standard library](https://gist.github.com/maustinstar/85528ce3dcd0b78fb55705cb0e031d94#file-misdiagnosis-swift-L15).

Checking in with @gottesmm to see if there is an easier way to actually identify this case?

Reply:
The reason this case is never actually identified as a mutation of the variable is because `Builtin.convertUnownedUnsafeToGuaranteed` uses `begin_access [read]`, instead of `begin_access [modify]`, and is not immediately clear that the value is later used as an inout.

Reply:
#39254 should remove the need for this function entirely.

Reply:
It appears we may need to keep this dedicated check, however I will explore if there is a way to optimize this a bit more.